provider/awssd/aws_sd.go

```diff
@@ -682,3 +1177,113 @@ func (p *AWSSDProvider) isAWSLoadBalancer(hostname string) bool {
 
 	return matchElb || matchNlb
 }
+
+// determine if a given service is of type A+SRV
+func (p *AWSSDProvider) isAAndSRVService(srv *sd.Service) bool {
+	return len(srv.DnsConfig.DnsRecords) == 2 &&
+		((aws.StringValue(srv.DnsConfig.DnsRecords[0].Type) == sd.RecordTypeA && aws.StringValue(srv.DnsConfig.DnsRecords[1].Type) == sd.RecordTypeSrv) ||
+			(aws.StringValue(srv.DnsConfig.DnsRecords[0].Type) == sd.RecordTypeSrv && aws.StringValue(srv.DnsConfig.DnsRecords[1].Type) == sd.RecordTypeA))
+}
+
+// returns the main DnsRecord of a service. In case of A+SRV service, it is the A DnsRecord
+func (p *AWSSDProvider) mainDnsRecord(srv *sd.Service) *sd.DnsRecord {
+	mainDnsRecord := srv.DnsConfig.DnsRecords[0]
+	// Cloud Map API doesn't guarantee a specific order for DnsRecords in a service. if "SRV" appears before "A", then make sure "A" is still the main DnsRecord.
+	if p.isAAndSRVService(srv) && aws.StringValue(mainDnsRecord.Type) == sd.RecordTypeSrv {
+		mainDnsRecord = srv.DnsConfig.DnsRecords[1]
+	}
+	return mainDnsRecord
+}
+
+func (p *AWSSDProvider) srvHostTargetSplit(target string) (port string, host string, prio string, weight string, err error) {
+	if !p.isSrvHostTarget(target) {
+		err = fmt.Errorf("endpoint target %s is not an host-based SRV target", target)
+		return
+	}
+	parts := strings.Split(target, " ")
+	port = parts[2]
+	host = parts[3]
+	prio = parts[0]
+	weight = parts[1]
+	return
+}
+
+func (p *AWSSDProvider) preferredSRVCombine(hostname string, ttl endpoint.TTL, target string) (prefSRV string, err error) {
```

Comment:
I struggle to understand the keyword `preferred`. What is the meaning in this context?

Reply:
it's the SRV endpoint that corresponds to the A endpoint. I replaced the keyword "preferred" with "corresponding"